File: cli/src/commands/resolve.ts
/**
 * mrsf resolve — Resolve (or unresolve) a comment.
 */

import type { Command } from "commander";
import chalk from "chalk";
import path from "node:path";
import { parseSidecar } from "../lib/parser.js";
import { writeSidecar } from "../lib/writer.js";
import { resolveComment, unresolveComment } from "../lib/comments.js";

export function registerResolve(program: Command): void {
  program
    .command("resolve <sidecar> <id>")
    .description("Resolve a comment by ID")
    .option("--cascade", "Also resolve direct replies")
    .option("-u, --undo", "Unresolve instead")
    .action(
      async (
        sidecar: string,
        id: string,
        opts: { cascade?: boolean; undo?: boolean },
      ) => {
        const parentOpts = program.opts();
        const cwd = parentOpts.cwd ?? process.cwd();
        const quiet = parentOpts.quiet ?? false;
        const sp = path.resolve(cwd, sidecar);

        const doc = await parseSidecar(sp);

        const ok = opts.undo
          ? unresolveComment(doc, id)
          : resolveComment(doc, id, opts.cascade);

        if (!ok) {
          console.error(chalk.red(`Comment ${id} not found in ${sidecar}.`));
          process.exit(1);
        }

        await writeSidecar(sp, doc);

        if (!quiet) {
          const action = opts.undo ? "Unresolved" : "Resolved";
          console.log(chalk.green(`${action} ${id} in ${sidecar}.`));
        }
      },
    );
}

File: python/src/mrsf/cli/resolve.py
"""mrsf resolve <sidecar> <id> — resolve or unresolve a comment."""

from __future__ import annotations

import sys

import click

from ..comments import resolve_comment, unresolve_comment
from ..parser import parse_sidecar
from ..writer import write_sidecar
from .main import MrsfContext, pass_ctx


@click.command("resolve")
@click.argument("sidecar", type=click.Path(exists=True))
@click.argument("comment_id")
@click.option("--cascade", is_flag=True, help="Also resolve direct replies.")
@click.option("-u", "--undo", is_flag=True, help="Unresolve the comment.")
@pass_ctx
def resolve_cmd(ctx: MrsfContext, sidecar: str, comment_id: str, cascade: bool, undo: bool) -> None:
    """Resolve or unresolve a comment by ID."""
    doc = parse_sidecar(sidecar)

    if undo:
        found = unresolve_comment(doc, comment_id)
        action = "Unresolved"
    else:
        found = resolve_comment(doc, comment_id, cascade=cascade)
        action = "Resolved"

    if not found:
        click.echo(click.style(f"Comment {comment_id} not found in {sidecar}", fg="red"), err=True)
        sys.exit(1)

    write_sidecar(sidecar, doc)
    if not ctx.quiet:
        click.echo(f"{action} {comment_id} in {sidecar}")
